src/main/java/org/codehaus/groovy/ast/tools/GenericsUtils.java

```diff
@@ -855,10 +855,16 @@ private static Map<GenericsType, GenericsType> doMakeDeclaringAndActualGenericsT
             GenericsType[] targetGenericsTypes = parameterizedType.redirect().getGenericsTypes();
             if (targetGenericsTypes != null) {
                 GenericsType[] sourceGenericsTypes = parameterizedType.getGenericsTypes();
-                if (sourceGenericsTypes == null) sourceGenericsTypes = EMPTY_GENERICS_ARRAY;
                 map = new LinkedHashMap<>();
-                for (int i = 0, m = sourceGenericsTypes.length, n = targetGenericsTypes.length; i < n; i += 1) {
-                    map.put(targetGenericsTypes[i], i < m ? sourceGenericsTypes[i] : targetGenericsTypes[i]);
+                if (sourceGenericsTypes == null) {
+                    //We have no sourceGenerics, so we need to return the erasures for the raw types (Groovy-10813)
+                    for (GenericsType targetGenericsType : targetGenericsTypes) {
+                        map.put(targetGenericsType, ClassHelper.OBJECT_TYPE.asGenericsType());
```

Comment:
The erasure type is not always Object.  Can you create an example where it is another type?

Reply:
I will try to find an example.

Reply:
I've played around with generics in java, but couldn't find an example where the erasure isn't of type Object.
I've also tried something like this:
```java
import java.util.function.*
interface T<E> extends Function<String, E>{}
class TT implements T{
  @Override
  Object apply(Object o) {
    return null
  }
}
```
and it still loses all type information. Maybe I'm just not advanced enough into generics and how they work. Could you perhaps help me here or give me a resource where to learn more about this? I've found some information about the topic [here](https://docs.oracle.com/javase/tutorial/java/generics/rawTypes.html) and [here](https://docs.oracle.com/javase/tutorial/java/generics/erasure.html)

Reply:
Another example where the type parameter has an upper bound: 
´´´java
interface A<T extends String> extends Function<T, T> {}
class B implements A{
    @Override
    Object apply(Object input) {
        return null
    }
}
´´´

Reply:
Uh, now I got an example, that has not Object as its type:
```java
interface A<T extends String> {
    T apply(T input);
}
class B implements A{
    @Override
    String apply(String input) {
        return null
    }
}
```

Reply:
I've added some tests that have an upper bound. I was a bit suprised that `T extends String & Object` just selects the first one when you use raw types, so the result is different from `T extends Object & String`